Reusable object templates

thabetx opened the pull request:
This pull request is about the reusable object templates project in GSoC 2017, addresses #70 .

The entry point for testing is selecting some objects and choosing "Save as a template group", the selected objects will be saved to a file, then the file will be immediately read and the sizes of the saved templates will be printed, this is just for testing purposes.

When saving tile objects, the tileset info will be saved as well, this encourages having standalone template groups that can be used across multiple objects, although for now, when reading the file, the tilesets are ignored.

The next step will be creating a template group document that will allow loading the template group into tiled.

Code wise, I've followed tileset format and map format for creating the new classes, the template group is till now similar to object group, maybe I will need to create a new class for a template instead of using the object class.

Reading and writing reuses some of the available code, I had to extract some of the object reading and writing code (the part that handles different types of objects), otherwise things are pretty similar to writing object groups albeit with some changes like not writing the coordinates of the object or its ID as they won't be relevant when creating new copies of the template.

So currently this hypothetical use case is simple and some parts of the code are not complete, I think having the template group ready and loading it into Tiled will require doing more checking and will reveal more cases that require better handling.

thabetx commented on src/libtiled/mapreader.cpp:
```diff
@@ -834,6 +873,48 @@ void MapReaderPrivate::readImageLayerImage(ImageLayer &imageLayer)
+
+    MapObject *object = new MapObject(name, type, {0, 0}, size);
+
+    readObjectData(object);
```
Alright, I didn't understand what you meant, I thought you referred to the template group as template.

So just to be clear the format will look like this
```
<templategroup ...>
 <template ...>
  <object ...>
    ...
  </object ...>
 </template ...>

 <template ...>
  <object ...>
    ...
  </object ...>
 </template>
</templategroup>
```
This looks more effective, I will update it and the template class makes more sense now.

And for writing, is there a better way other than creating a new function just to prevent writing the id and the coordinates? Does sending a parameter to tell it whether to write them or not be appropriate?

bjorn replied:
I think it would be fine to prevent the writing of the ID by just making sure it is 0 and adding a check for this in `MapWriterPrivate::writeObject`. For now, I would not bother with preventing the writing of the position.

thabetx replied:
Okay that would allow more code reuse, but what do you mean by setting id to 0, wouldn't that change the original object?

bjorn replied:
Well, only the object owned by the template, which should be a copy, would have its ID set to 0.

thabetx replied:
Okay right, thanks!

thabetx replied:
`template` is a keyword in C++, would `objectTemplate` as the name of the class be fine?

thabetx commented:
I've update the format, added the tileset references to the templategroup and fixed some review points.

I'm working on changing the writing function to work directly on a template group instead of an object list, and will add more information to the template group and the templates.

thabetx commented:
I updated the saving function and will start working on the `TemplateGroupDocument`.

I added a name to `templateGroup` and a name and ID to `objectTemplate`, not sure yet how the ID will be used.

To prevent writing the ID and the coordinates of the `mapObject` in the `objectTemplate`, I used the suggestion of setting the ID of the cloned `mapObject` to 0 and used it as a check.

thabetx commented:
I implemented the `TemplateGroupDocument` functions and now it's used for testing. I've also squashed some of the commits, testing is still done by selecting some objects then choosing "save as template group".

bjorn commented on src/tiled/templategroupdocument.cpp:
```diff
@@ -0,0 +1,96 @@
+{
+    QString displayName = mTemplateGroup->name();
+
+    if(displayName.isEmpty())
```
Coding style: please put a space after `if`.

bjorn commented on src/tiled/tmxmapformat.cpp:
```diff
@@ -166,3 +166,56 @@ bool TsxTilesetFormat::supportsFile(const QString &fileName) const
+    mError.clear();
+
+    MapReader reader;
+    TemplateGroup *templateGroup =  reader.readTemplateGroup(fileName);
```
Coding style: one space too many after `=`.

bjorn commented on src/tiled/mapdocument.cpp:
```diff
@@ -125,6 +127,58 @@ MapDocument::~MapDocument()
+/**
+ * Currently, this is the entry point for testing templates,
+ * it creates a templateGroup from the selected objects except the last object,
+ * then creates a templatGroupDocument from the templateGroup,
```
Typo: templatGroupDocument (also, I think it would be better to match the casing with the class name).

bjorn commented on src/tiled/mapdocument.cpp:
```diff
@@ -125,6 +127,58 @@ MapDocument::~MapDocument()
+
+    QString fileName = QLatin1String("templateGroup.ttx");
+
+    TemplateGroupDocument *templateGroupDocument = new TemplateGroupDocument(templateGroup, fileName);
```
No change required, but just to note, that to have less repetition I am often using `auto` for such lines these days, since the type is completely obvious:
```c++
auto templateGroupDocument = new TemplateGroupDocument(templateGroup, fileName);
```
